### PR TITLE
update draining and setup_marathon_job docs for #529

### PR DESCRIPTION
**Pull request by huadongliu**

### CODE REVIEW

File: docs/source/deploy_marathon_services.rst [hunk 1]
```diff
@@ -2,6 +2,20 @@ deploy_marathon_services (bash script)
...
+The maximum number of service.instances that ``setup_marathon_job`` takes,
+and the maximum number of parallel deployment processes are configurable
+in puppet. By default, ``setup_marathon_job`` takes one service.instance
+each time and there are at most five parallel ``setup_marathon_job`` processes.
+
+With below config in ``modules/profile_paasta/data/clusters/mesosstage.yaml``,
+each ``setup_marathon_job`` in the cluster mesosstage can take 64 service.instances and xargs can fork four
+``setup_marathon_job`` process to deploy services in parallel::
+
+  profile_paasta::marathon::max_serviceinstances_per_deployment: 64
```
Comment (solarkennedy) [inline docs/source/deploy_marathon_services.rst]: Don't include these here, they are yelp-specific.

File: docs/source/setup_marathon_job.rst [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+``setup_marathon_job`` grabs a zookeeper lock for each app while it is
+being deployed. Locking is not necessary if `deploy_marathon_services <deploy_marathon_services.html>`_
+is the only source of service deployment. It provides safety against another
+``setup_marathon_job`` being launched, e.g. manually by a system admin.
+
+It takes multiple ``setup_marathon_job`` runs to advance an app from its
+current state to the desired state. For example, below steps occur in
+sequential ``setup_marathon_job`` runs in a crossover bounce:
+
+* The new app is created with marathon and marathon is working to launch
```
Comment (solarkennedy) [inline docs/source/setup_marathon_job.rst]: I wouldn't duplicate these docs here, instead link to the existing bounce documentation

File: docs/source/workflow.rst [hunk 3]
```diff
@@ -231,6 +232,37 @@ Valid options are:
...
+  used at Yelp to provide APIs to query and change state of a service instance.
+  The hacheck draining method requests hacheck to mark down an instance. HAProxy
+  checks with hacheck periodically to keep its view of instance state up-to-date.
+  The hacheck draining method will wait for a configurable ``delay`` to make sure
+  HAProxy has the update before considering safe to kill an instance. Note that
+  the hacheck draining method sets an expiration when marking an instance down on
+  hacheck. hacheck will drop the down state if it receives a status query after
+  expiration.
+
+**What is the rationale behind the expiration logic? We have seen boucing being
```
Comment (solarkennedy) [inline docs/source/workflow.rst]: This formatting looks odd., is it meant to be indented with the hacheck description?

Also please don't make it all bold

Comment (huadongliu) [inline docs/source/workflow.rst]: The bold part is actually a question for reviewers regarding the draining expiration logic.

File: docs/source/setup_marathon_job.rst (hunk 2 above)
Comment (huadongliu) [inline docs/source/setup_marathon_job.rst]: Line 114 has a link to setup_marathon_job. I thought it is the intended place for setup_marathon_job docs.

114 a thin wrapper around `setup_marathon_job <setup_marathon_job.html>`_.

File: docs/source/workflow.rst (hunk 3 above)
Comment (solarkennedy) [inline docs/source/workflow.rst]: It is mostly a safety feature. Things can go wrong for all sorts of reasons, and adding the expiration ensures that is something really goes wrong or the bounce code stops running or starts crashing or whatever, any load balancer changes will "undo" themselves and get traffic flowing again.

File: docs/source/workflow.rst [hunk 4]
```diff
@@ -111,11 +111,43 @@ A yelpsoa-configs master runs `generate_deployments_for_service <generated/paast
 frequently. The generated ``deployments.json`` appears in ``/nail/etc/services/service_name`` throughout the cluster.
 
 Marathon masters run `deploy_marathon_services <deploy_marathon_services.html>`_,
-a thin wrapper around `setup_marathon_job <setup_marathon_job.html>`_.
+a thin wrapper around ``setup_marathon_job``.
 These scripts parse ``deployments.json`` and the current cluster state,
 then issue comands to Marathon to put the cluster into the right state
 -- cluster X should be running version Y of service Z.
 
+``setup_marathon_job`` iterates over a list of service.instances to avoid
```
Comment (solarkennedy) [inline docs/source/workflow.rst]: This documentation is too low level to be in the "workflow" section.

Users don't really need to worry about these internals.

Please make a new section in the bounce docs for these and move them there. Day-to-day users don't need to know about these details

**Pull request merged**